File: tools/sdks/compose/src/main/kotlin/io/thunder/compose/components/presentation/auth/SignUp.kt
package io.thunder.compose.components.presentation.auth

import androidx.compose.foundation.layout.*
import androidx.compose.foundation.text.BasicText
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.runtime.*
import androidx.compose.ui.Modifier
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.unit.dp
import io.thunder.android.*
import io.thunder.compose.LocalThunder
import io.thunder.compose.ThunderState
import io.thunder.compose.components.actions.BaseThunderSignUpButton
import kotlinx.coroutines.launch

/** State passed to the [BaseThunderSignUp] builder slot. */
@Stable
class ThunderSignUpState {
    var inputs by mutableStateOf<List<FlowInput>>(emptyList())
        internal set
    var actions by mutableStateOf<List<FlowAction>>(emptyList())
        internal set
    var isLoading by mutableStateOf(false)
        internal set
    var error by mutableStateOf<String?>(null)
        internal set

    internal var flowId: String? = null
    internal var onSubmit: (String) -> Unit = {}
    private val fieldValues = mutableStateMapOf<String, String>()

    fun fieldValue(name: String): String = fieldValues[name] ?: ""
    fun setField(name: String, value: String) { fieldValues[name] = value }
    fun fields(): Map<String, String> = fieldValues.toMap()
    fun submit(actionId: String) = onSubmit(actionId)

    internal fun update(response: EmbeddedFlowResponse) {
        flowId = response.flowId
        inputs = response.data?.inputs ?: emptyList()
        actions = response.data?.actions ?: emptyList()
    }
}

/** App-native sign-up form (spec §8.4 Presentation). */
@Composable
fun ThunderSignUp(
    modifier: Modifier = Modifier,
    onComplete: (() -> Unit)? = null,
    onError: ((String) -> Unit)? = null,
) {
    val thunderState = LocalThunder.current
    val i18n = thunderState.i18n
    BaseThunderSignUp(modifier = modifier, onComplete = onComplete, onError = onError) { state ->
        Column(modifier = Modifier.padding(16.dp), verticalArrangement = Arrangement.spacedBy(12.dp)) {
            BasicText(i18n.resolve("signUp.title"))
            state.error?.let { BasicText(it) }
            state.inputs.forEach { input ->
                BasicTextField(
                    value = state.fieldValue(input.name),
                    onValueChange = { state.setField(input.name, it) },
                    modifier = Modifier.fillMaxWidth().defaultMinSize(minHeight = 44.dp)
                        .semantics { contentDescription = input.name },
                )
            }
            state.actions.forEach { action ->
                BaseThunderSignUpButton(label = action.label ?: i18n.resolve("signUp.submit")) {
                    state.submit(action.id)
                }
            }
            if (state.isLoading) BasicText(i18n.resolve("signUp.loading"))
        }
    }
}

/** Unstyled base variant (spec §8.3). */
@Composable
fun BaseThunderSignUp(
    modifier: Modifier = Modifier,
    onComplete: (() -> Unit)? = null,
    onError: ((String) -> Unit)? = null,
    content: @Composable (ThunderSignUpState) -> Unit,
) {
    val thunderState = LocalThunder.current
    val scope = rememberCoroutineScope()
    val signUpState = remember { ThunderSignUpState() }

    signUpState.onSubmit = { actionId ->
        scope.launch {
            signUpState.isLoading = true; signUpState.error = null
            try {
                val payload = EmbeddedSignInPayload(flowId = signUpState.flowId, actionId = actionId, inputs = signUpState.fields())
                val response = thunderState.client.signUp(payload = payload)
                handleSignUpResponse(response, signUpState, thunderState, onComplete, onError)
            } catch (e: Exception) {
                signUpState.error = e.message
                onError?.invoke(e.message ?: "Sign-up failed")
            } finally { signUpState.isLoading = false }
        }
    }

    LaunchedEffect(Unit) {
        signUpState.isLoading = true
        try {
            val response = thunderState.client.signUp()
            handleSignUpResponse(response, signUpState, thunderState, onComplete, onError)
        } catch (e: Exception) {
            signUpState.error = e.message
            onError?.invoke(e.message ?: "Sign-up failed")
        } finally { signUpState.isLoading = false }
    }

    Box(modifier = modifier) { content(signUpState) }
}

private suspend fun handleSignUpResponse(
    response: EmbeddedFlowResponse,
    state: ThunderSignUpState,
    thunderState: ThunderState,
    onComplete: (() -> Unit)?,
    onError: ((String) -> Unit)?,
) {
    when (response.flowStatus) {
        FlowStatus.COMPLETE -> { thunderState.refresh(); onComplete?.invoke() }
        FlowStatus.PROMPT_ONLY -> state.update(response)
        FlowStatus.ERROR -> {
            val msg = response.failureReason ?: "Sign-up failed"
            state.error = msg
            onError?.invoke(msg)
        }
    }
}

File: tools/sdks/swiftui/Sources/ThunderIDSwiftUI/components/presentation/auth/SignUp.swift
import SwiftUI
import ThunderID

/// App-native sign-up form. Drives the Flow Execution API registration loop (spec §8.4 Presentation).
public struct ThunderSignUp: View {
    @EnvironmentObject private var state: ThunderState
    @EnvironmentObject private var i18n: ThunderI18n
    public let applicationId: String
    public let onComplete: (() -> Void)?
    public let onError: ((String) -> Void)?

    public init(
        applicationId: String,
        onComplete: (() -> Void)? = nil,
        onError: ((String) -> Void)? = nil
    ) {
        self.applicationId = applicationId
        self.onComplete = onComplete
        self.onError = onError
    }

    public var body: some View {
        BaseThunderSignUp(applicationId: applicationId, onComplete: onComplete, onError: onError) { signUpState in
            VStack(alignment: .leading, spacing: 12) {
                Text(i18n.resolve("signUp.title"))
                    .accessibilityAddTraits(.isHeader)
                if let error = signUpState.error {
                    Text(error).foregroundColor(.red)
                }
                ForEach(signUpState.inputs, id: \.name) { input in
                    if input.type == "password" {
                        SecureField(input.name, text: signUpState.binding(for: input.name))
                            .accessibilityLabel(input.name)
                            .frame(minHeight: 44)
                    } else {
                        TextField(input.name, text: signUpState.binding(for: input.name))
                            .accessibilityLabel(input.name)
                            .frame(minHeight: 44)
                    }
                }
                ForEach(signUpState.actions, id: \.id) { action in
                    Button(action.label ?? i18n.resolve("signUp.submit")) {
                        signUpState.submit(actionId: action.id)
                    }
                    .disabled(signUpState.isLoading)
                    .frame(minWidth: 44, minHeight: 44)
                }
                if signUpState.isLoading {
                    Text(i18n.resolve("signUp.loading"))
                }
            }
            .padding()
        }
    }
}

/// Mutable state passed to the BaseThunderSignUp builder.
@MainActor
public final class ThunderSignUpState: ObservableObject {
    @Published public private(set) var inputs: [FlowInput] = []
    @Published public private(set) var actions: [FlowAction] = []
    @Published public private(set) var isLoading: Bool = false
    @Published public private(set) var error: String?

    private var fieldValues: [String: String] = [:]
    private var flowId: String?
    let submitAction: (String, [String: String], String?) -> Void

    init(submit: @escaping (String, [String: String], String?) -> Void) {
        self.submitAction = submit
    }

    public func binding(for name: String) -> Binding<String> {
        Binding(get: { self.fieldValues[name] ?? "" }, set: { self.fieldValues[name] = $0 })
    }

    public func submit(actionId: String) { submitAction(actionId, fieldValues, flowId) }

    func update(from response: EmbeddedFlowResponse) {
        flowId = response.flowId
        inputs = response.data?.inputs ?? []
        actions = response.data?.actions ?? []
    }
}

/// Unstyled base variant (spec §8.3).
public struct BaseThunderSignUp<Content: View>: View {
    @EnvironmentObject private var state: ThunderState
    public let applicationId: String
    public let onComplete: (() -> Void)?
    public let onError: ((String) -> Void)?
    public let content: (ThunderSignUpState) -> Content

    @StateObject private var signUpState = ThunderSignUpState(submit: { _, _, _ in })

    public init(
        applicationId: String,
        onComplete: (() -> Void)? = nil,
        onError: ((String) -> Void)? = nil,
        @ViewBuilder content: @escaping (ThunderSignUpState) -> Content
    ) {
        self.applicationId = applicationId
        self.onComplete = onComplete
        self.onError = onError
        self.content = content
    }

    public var body: some View {
        content(signUpState)
            .task { await initFlow() }
    }

    private func initFlow() async {
        signUpState.isLoading = true
        defer { signUpState.isLoading = false }
        do {
            let response = try await state.client.signUp()
            await handleResponse(response)
        } catch {
            signUpState.error = error.localizedDescription
            onError?(error.localizedDescription)
        }
    }

    private func submit(actionId: String, inputs: [String: String], flowId: String?) async {
        signUpState.isLoading = true
        defer { signUpState.isLoading = false }
        do {
            let payload = EmbeddedSignInPayload(flowId: flowId, actionId: actionId, inputs: inputs)
            let response = try await state.client.signUp(payload: payload)
            await handleResponse(response)
        } catch {
            signUpState.error = error.localizedDescription
            onError?(error.localizedDescription)
        }
    }

    private func handleResponse(_ response: EmbeddedFlowResponse) async {
        switch response.flowStatus {
        case .complete:
            await state.refresh()
            onComplete?()
        case .promptOnly:
            signUpState.update(from: response)
        case .error:
            let msg = response.failureReason ?? "Sign-up failed"
            signUpState.error = msg
            onError?(msg)
        }
    }
}
